### ISSUE TITLE
Integrate company with `indent-for-tab-command`

### ISSUE THREAD

**Issue by bbatsov** [company-mode/company-mode#94]
For some reason only "RET" does completion and "TAB" behaves pretty much as it always did when company completion popup is visible. 

I think this is related to me having set `tab-always-indent` to `'complete`. The issue can be reproduced easily with a clean install of Prelude (I'm tired of users always asking for `auto-complete-mode`, so I decided to integrate a better solution and pray they won't notice the difference). I tried setting `tab-always-indent` to its default value `t` and the extra completion buffer I got stopped appearing (as expected), but "TAB" still doesn't complete with the selected entry in the company popup buffer.

**Comment by jboadas** [comment]
This work great, thanks
But I have to write this way to work for me.
(with-eval-after-load 'company
  (define-key company-mode-map [remap indent-for-tab-command]
    ' company-indent-for-tab-command))

**Comment by izahn** [comment]
Just a note for any future wanderers of the internet that we can now bind `company-indent-for-tab-command` to tab. Should work better than the workaround posted above.

**Comment by ryukinix** [comment]
@izahn but where is defined `company-indent-for-tab-command`? I didn't found.

**Comment by mnewt** [comment]
If you want to invoke completion with `tab` then I think that's the correct command. This is how I'm using it and it seems to work well (it wraps the normal tab commands):

```
(global-company-mode)
(global-set-key (kbd "TAB") #'company-indent-or-complete-common)
```

**Comment by paxperscientiam** [comment]
@mnewt signed in just to give you a "hooray" 👍

**Comment by didibus** [comment]
`company-indent-or-complete-common` doesn't take the same argument as `indent-for-tab-command` which means I can't use it as a fallback for `adjust-parens-fallback-dedent-function` when using adjust-parens, as it assumes to fallback to a function of similar interface as `indent-for-tab-command`. Is there no similar function for company that takes the interface of `indent-for-tab-command` ?

**Comment by dgutov** [comment]
How would it handle the ARG argument?

If it'll simply ignore it, you can write a wrapper function very easily.

**Comment by didibus** [comment]
That's what I did, but I'm assuming indent-for-tab-command does something with the arg, otherwise why does it take one? Not sure if it makes sense for company-indent-or-complete to also do something with the arg or not.

**Comment by dgutov** [comment]
So what's the behavior that `adjust-parens-fallback-dedent-function` expects? And what package does it come from? I don't see it in the stock Emacs.

**Comment by didibus** [comment]
It seems `indent-for-tab-command`, in base Emacs takes an arg that is used to pass in a prefix, and when called that way, it will indent the whole s-exp as well, by calling indent-rigidly instead.

See code here: https://github.com/emacs-mirror/emacs/blob/master/lisp/indent.el#L162

`adjust-parens-fallback-dedent-function` is used by adjust-parens. Basically, adjust-parents takes over TAB.  But in a lot of cases, it can simply delegate back to the original TAB command, which is `indent-for-tab-command`. Thus it calls the function defined in `adjust-parens-fallback-dedent-function` which by default is `indent-for-tab-command`.

But when replaced with `company-indent-or-complete` it fails, because that doesn't have the same interface (takes no args) as `indent-for-tab-command`.

My conclusion is that, it seems there should be a `company-indent-or-complete-command` that is also an (interactive "P") command which takes a prefix arg and indents rigidly if so. That way it would be true to the default `indent-for-tab-command` behavior.

**Comment by dgutov** [comment]
Will this work for your purposes?

```diff
diff --git a/company.el b/company.el
index e05f668..df01c96 100644
--- a/company.el
+++ b/company.el
@@ -2171,9 +2171,9 @@ With ARG, move by that many elements."
               (current-prefix-arg arg))
           (call-interactively 'company-select-next))))))
 
-(defun company-indent-or-complete-common ()
+(defun company-indent-or-complete-common (arg)
   "Indent the current line or region, or complete the common part."
-  (interactive)
+  (interactive "P")
   (cond
    ((use-region-p)
     (indent-region (region-beginning) (region-end)))
@@ -2183,7 +2183,7 @@ With ARG, move by that many elements."
    ((let ((old-point (point))
           (old-tick (buffer-chars-modified-tick))
           (tab-always-indent t))
-      (call-interactively #'indent-for-tab-command)
+      (indent-for-tab-command arg)
       (when (and (eq old-point (point))
                  (eq old-tick (buffer-chars-modified-tick)))
         (company-complete-common))))))
```

**Comment by didibus** [comment]
Looks like it would ya!

**Comment by dgutov** [comment]
Any chance you can test it?22:44, January 7, 2020, "Didier A." <notifications@github.com>:Looks like it would ya!

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.
-- Dmitry Gutov

**Comment by didibus** [comment]
Ok, I tested it, and it does work. It both now is a proper replacement for indent-tab-for-command from adjust-parens, as well as you can now use a prefix arg to do an indent-rigidly as with indent-for-tab even when company-indent-or-complete-common is set as the TAB command.

**Comment by dgutov** [comment]
Thanks for testing, pushed.

**Comment by andreyorst** [comment]
It seems that current implementation of `company-indent-or-complete-common` ignores value of `tab-first-completion` variable by setting `tab-always-indent` to `t` resulting in completion working in front of parentheses even when `tab-first-completion` is set to `word-or-paren-or-punct` with `indent-for-tab-command` remapped to `company-indent-or-complete-common`

**Comment by dgutov** [comment]
That's true. I never knew about this variable, or really set out to make a fully compatible replacement for `indent-for-tab-command`.

But patches welcome.

**Comment by dgutov** [comment]
Included in https://github.com/company-mode/company-mode/pull/1534 (hooking into `tab-always-indent` without having to set up binding).
